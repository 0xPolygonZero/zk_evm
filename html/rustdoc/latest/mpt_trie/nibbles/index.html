<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="Define `Nibbles` and how to convert bytes, hex prefix encodings and strings into nibbles."><title>mpt_trie::nibbles - Rust</title><script>if(window.location.protocol!=="file:")document.head.insertAdjacentHTML("beforeend","SourceSerif4-Regular-46f98efaafac5295.ttf.woff2,FiraSans-Regular-018c141bf0843ffd.woff2,FiraSans-Medium-8f9a781e4970d388.woff2,SourceCodePro-Regular-562dcc5011b6de7d.ttf.woff2,SourceCodePro-Semibold-d899c5a5c4aeb14a.ttf.woff2".split(",").map(f=>`<link rel="preload" as="font" type="font/woff2" crossorigin href="../../static.files/${f}">`).join(""))</script><link rel="stylesheet" href="../../static.files/normalize-76eba96aa4d2e634.css"><link rel="stylesheet" href="../../static.files/rustdoc-492a78a4a87dcc01.css"><meta name="rustdoc-vars" data-root-path="../../" data-static-root-path="../../static.files/" data-current-crate="mpt_trie" data-themes="" data-resource-suffix="" data-rustdoc-version="1.82.0-nightly (1f12b9b0f 2024-08-27)" data-channel="nightly" data-search-js="search-acd2e111c9cb27e3.js" data-settings-js="settings-4313503d2e1961c2.js" ><script src="../../static.files/storage-118b08c4c78b968e.js"></script><script defer src="../sidebar-items.js"></script><script defer src="../../static.files/main-cb0df477c2d67d00.js"></script><noscript><link rel="stylesheet" href="../../static.files/noscript-3b12f09e550e0385.css"></noscript><link rel="alternate icon" type="image/png" href="../../static.files/favicon-32x32-422f7d1d52889060.png"><link rel="icon" type="image/svg+xml" href="../../static.files/favicon-2c020d218678b618.svg"><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/katex@0.11.1/dist/katex.min.css"
    integrity="sha384-zB1R0rpPzHqg7Kpt0Aljp8JPLqbXI3bhnPWROx27a9N0Ll6ZP/+DiW/UqRcLbRjq" crossorigin="anonymous">
<script defer src="https://cdn.jsdelivr.net/npm/katex@0.11.1/dist/katex.min.js"
    integrity="sha384-y23I5Q6l+B6vatafAwxRu/0oK/79VlbSz7Q9aiSZUvyWYIYsd+qj+o24G5ZU2zJz"
    crossorigin="anonymous"></script>
<script defer src="https://cdn.jsdelivr.net/npm/katex@0.11.1/dist/contrib/auto-render.min.js"
    integrity="sha384-kWPLUVMOks5AQFrykwIup5lo0m3iMkkHrD0uJ4H5cjeGihAutqP0yW0J6dpFiVkI"
    crossorigin="anonymous"></script>
<script>
    document.addEventListener("DOMContentLoaded", function () {
        renderMathInElement(document.body, {
            fleqn: false,
            macros: {
                "\\F": "\\mathbb{F}",
                "\\G": "\\mathbb{G}",
                "\\O": "\\mathcal{O}",
                "\\(": "\\left(",
                "\\)": "\\right)",
                "\\norm": "\\left\\vert #1 \\right\\vert",
                "\\set": "\\mathcal{ #1 }",
            },
            delimiters: [
                { left: "$$", right: "$$", display: true },
                { left: "\\(", right: "\\)", display: false },
                { left: "$", right: "$", display: false },
                { left: "\\[", right: "\\]", display: true }
            ]
        });
    });
</script>
</head><body class="rustdoc mod"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="mobile-topbar"><button class="sidebar-menu-toggle" title="show sidebar"></button></nav><nav class="sidebar"><div class="sidebar-crate"><h2><a href="../../mpt_trie/index.html">mpt_<wbr>trie</a><span class="version">0.4.1</span></h2></div><h2 class="location"><a href="#">Module nibbles</a></h2><div class="sidebar-elems"><section><ul class="block"><li><a href="#structs">Structs</a></li><li><a href="#enums">Enums</a></li><li><a href="#traits">Traits</a></li><li><a href="#types">Type Aliases</a></li></ul></section><h2><a href="../index.html">In crate mpt_<wbr>trie</a></h2></div></nav><div class="sidebar-resizer"></div><main><div class="width-limiter"><rustdoc-search></rustdoc-search><section id="main-content" class="content"><div class="main-heading"><h1>Module <a href="../index.html">mpt_trie</a>::<wbr><a class="mod" href="#">nibbles</a><button id="copy-path" title="Copy item path to clipboard">Copy item path</button></h1><span class="out-of-band"><a class="src" href="../../src/mpt_trie/nibbles.rs.html#1-1705">source</a> · <button id="toggle-all-docs" title="collapse all docs">[<span>&#x2212;</span>]</button></span></div><details class="toggle top-doc" open><summary class="hideme"><span>Expand description</span></summary><div class="docblock"><p>Define <a href="struct.Nibbles.html" title="struct mpt_trie::nibbles::Nibbles"><code>Nibbles</code></a> and how to convert bytes, hex prefix encodings and
strings into nibbles.</p>
</div></details><h2 id="structs" class="section-header">Structs<a href="#structs" class="anchor">§</a></h2><ul class="item-table"><li><div class="item-name"><a class="struct" href="struct.Nibbles.html" title="struct mpt_trie::nibbles::Nibbles">Nibbles</a></div><div class="desc docblock-short">A sequence of nibbles which is used as the key type into
<a href="../partial_trie/trait.PartialTrie.html" title="trait mpt_trie::partial_trie::PartialTrie"><code>PartialTrie</code></a>.</div></li><li><div class="item-name"><a class="struct" href="struct.NibblesIntern.html" title="struct mpt_trie::nibbles::NibblesIntern">Nibbles<wbr>Intern</a></div><div class="desc docblock-short">Little-endian large integer type
Used for the internal representation of a sequence of nibbles.
The choice of [u64; 5] accommodates the 260-bit key requirement efficiently by
leveraging 64-bit instructions for performance, while providing an additional u64
to handle the overflow case beyond the 256-bit capacity of [u64; 4].</div></li><li><div class="item-name"><a class="struct" href="struct.StrToNibblesError.html" title="struct mpt_trie::nibbles::StrToNibblesError">StrTo<wbr>Nibbles<wbr>Error</a></div><div class="desc docblock-short">An error encountered when converting a string to a sequence of nibbles.</div></li></ul><h2 id="enums" class="section-header">Enums<a href="#enums" class="anchor">§</a></h2><ul class="item-table"><li><div class="item-name"><a class="enum" href="enum.BytesToNibblesError.html" title="enum mpt_trie::nibbles::BytesToNibblesError">Bytes<wbr>ToNibbles<wbr>Error</a></div><div class="desc docblock-short">Errors encountered when converting from <code>Bytes</code> to <code>Nibbles</code>.</div></li><li><div class="item-name"><a class="enum" href="enum.FromHexPrefixError.html" title="enum mpt_trie::nibbles::FromHexPrefixError">From<wbr>HexPrefix<wbr>Error</a></div><div class="desc docblock-short">Errors encountered when converting to hex prefix encoding to nibbles.</div></li><li><div class="item-name"><a class="enum" href="enum.NibblesToTypeError.html" title="enum mpt_trie::nibbles::NibblesToTypeError">Nibbles<wbr>ToType<wbr>Error</a></div><div class="desc docblock-short">Error type for conversion.</div></li></ul><h2 id="traits" class="section-header">Traits<a href="#traits" class="anchor">§</a></h2><ul class="item-table"><li><div class="item-name"><a class="trait" href="trait.ToNibbles.html" title="trait mpt_trie::nibbles::ToNibbles">ToNibbles</a></div><div class="desc docblock-short">Because there are two different ways to convert to <code>Nibbles</code>, we don’t want
to rely on <code>From</code>. Instead, we’ll define a new trait that defines both
conversions.</div></li></ul><h2 id="types" class="section-header">Type Aliases<a href="#types" class="anchor">§</a></h2><ul class="item-table"><li><div class="item-name"><a class="type" href="type.Nibble.html" title="type mpt_trie::nibbles::Nibble">Nibble</a></div><div class="desc docblock-short">A Nibble has 4 bits and is stored as <code>u8</code>.</div></li></ul></section></div></main></body></html>