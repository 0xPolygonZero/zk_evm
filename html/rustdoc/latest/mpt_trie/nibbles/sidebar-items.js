window.SIDEBAR_ITEMS = {"enum":["BytesToNibblesError","FromHexPrefixError","NibblesToTypeError"],"struct":["Nibbles","NibblesIntern","StrToNibblesError"],"trait":["ToNibbles"],"type":["Nibble"]};