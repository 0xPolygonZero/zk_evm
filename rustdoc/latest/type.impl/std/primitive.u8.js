(function() {
    var type_impls = Object.fromEntries([["mpt_trie",[["<details class=\"toggle implementors-toggle\" open><summary><section id=\"impl-ToNibbles-for-u8\" class=\"impl\"><a class=\"src rightside\" href=\"src/mpt_trie/nibbles.rs.html#183\">source</a><a href=\"#impl-ToNibbles-for-u8\" class=\"anchor\">§</a><h3 class=\"code-header\">impl <a class=\"trait\" href=\"mpt_trie/nibbles/trait.ToNibbles.html\" title=\"trait mpt_trie::nibbles::ToNibbles\">ToNibbles</a> for <a class=\"primitive\" href=\"https://doc.rust-lang.org/nightly/std/primitive.u8.html\">u8</a></h3></section></summary><div class=\"impl-items\"><details class=\"toggle method-toggle\" open><summary><section id=\"method.to_nibbles\" class=\"method trait-impl\"><a class=\"src rightside\" href=\"src/mpt_trie/nibbles.rs.html#183\">source</a><a href=\"#method.to_nibbles\" class=\"anchor\">§</a><h4 class=\"code-header\">fn <a href=\"mpt_trie/nibbles/trait.ToNibbles.html#tymethod.to_nibbles\" class=\"fn\">to_nibbles</a>(self) -&gt; <a class=\"struct\" href=\"mpt_trie/nibbles/struct.Nibbles.html\" title=\"struct mpt_trie::nibbles::Nibbles\">Nibbles</a></h4></section></summary><div class='docblock'>Convert the type to a sequence of nibbles. <a href=\"mpt_trie/nibbles/trait.ToNibbles.html#tymethod.to_nibbles\">Read more</a></div></details><details class=\"toggle method-toggle\" open><summary><section id=\"method.to_nibbles_byte_padded\" class=\"method trait-impl\"><a class=\"src rightside\" href=\"src/mpt_trie/nibbles.rs.html#64-72\">source</a><a href=\"#method.to_nibbles_byte_padded\" class=\"anchor\">§</a><h4 class=\"code-header\">fn <a href=\"mpt_trie/nibbles/trait.ToNibbles.html#method.to_nibbles_byte_padded\" class=\"fn\">to_nibbles_byte_padded</a>(self) -&gt; <a class=\"struct\" href=\"mpt_trie/nibbles/struct.Nibbles.html\" title=\"struct mpt_trie::nibbles::Nibbles\">Nibbles</a><div class=\"where\">where\n    Self: <a class=\"trait\" href=\"https://doc.rust-lang.org/nightly/core/marker/trait.Sized.html\" title=\"trait core::marker::Sized\">Sized</a>,</div></h4></section></summary><div class='docblock'>Convert the type to a sequence of nibbles but pad to the nearest byte.</div></details></div></details>","ToNibbles","mpt_trie::nibbles::Nibble"],["<details class=\"toggle implementors-toggle\" open><summary><section id=\"impl-TryFrom%3CNibblesIntern%3E-for-u8\" class=\"impl\"><a class=\"src rightside\" href=\"src/mpt_trie/nibbles.rs.html#33-39\">source</a><a href=\"#impl-TryFrom%3CNibblesIntern%3E-for-u8\" class=\"anchor\">§</a><h3 class=\"code-header\">impl <a class=\"trait\" href=\"https://doc.rust-lang.org/nightly/core/convert/trait.TryFrom.html\" title=\"trait core::convert::TryFrom\">TryFrom</a>&lt;<a class=\"struct\" href=\"mpt_trie/nibbles/struct.NibblesIntern.html\" title=\"struct mpt_trie::nibbles::NibblesIntern\">NibblesIntern</a>&gt; for <a class=\"primitive\" href=\"https://doc.rust-lang.org/nightly/std/primitive.u8.html\">u8</a></h3></section></summary><div class=\"impl-items\"><details class=\"toggle\" open><summary><section id=\"associatedtype.Error\" class=\"associatedtype trait-impl\"><a class=\"src rightside\" href=\"src/mpt_trie/nibbles.rs.html#33-39\">source</a><a href=\"#associatedtype.Error\" class=\"anchor\">§</a><h4 class=\"code-header\">type <a href=\"https://doc.rust-lang.org/nightly/core/convert/trait.TryFrom.html#associatedtype.Error\" class=\"associatedtype\">Error</a> = &amp;'static <a class=\"primitive\" href=\"https://doc.rust-lang.org/nightly/std/primitive.str.html\">str</a></h4></section></summary><div class='docblock'>The type returned in the event of a conversion error.</div></details><details class=\"toggle method-toggle\" open><summary><section id=\"method.try_from\" class=\"method trait-impl\"><a class=\"src rightside\" href=\"src/mpt_trie/nibbles.rs.html#33-39\">source</a><a href=\"#method.try_from\" class=\"anchor\">§</a><h4 class=\"code-header\">fn <a href=\"https://doc.rust-lang.org/nightly/core/convert/trait.TryFrom.html#tymethod.try_from\" class=\"fn\">try_from</a>(u: <a class=\"struct\" href=\"mpt_trie/nibbles/struct.NibblesIntern.html\" title=\"struct mpt_trie::nibbles::NibblesIntern\">NibblesIntern</a>) -&gt; <a class=\"enum\" href=\"https://doc.rust-lang.org/nightly/core/result/enum.Result.html\" title=\"enum core::result::Result\">Result</a>&lt;<a class=\"primitive\" href=\"https://doc.rust-lang.org/nightly/std/primitive.u8.html\">u8</a>, &amp;'static <a class=\"primitive\" href=\"https://doc.rust-lang.org/nightly/std/primitive.str.html\">str</a>&gt;</h4></section></summary><div class='docblock'>Performs the conversion.</div></details></div></details>","TryFrom<NibblesIntern>","mpt_trie::nibbles::Nibble"]]]]);
    if (window.register_type_impls) {
        window.register_type_impls(type_impls);
    } else {
        window.pending_type_impls = type_impls;
    }
})()
//{"start":55,"fragment_lengths":[4658]}