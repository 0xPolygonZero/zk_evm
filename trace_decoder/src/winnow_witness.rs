//! Sane parser for witness format.
//!
//! Based on [this specification](*https://gist.github.com/mandrigin/ff7eccf30d0ef9c572bafcb0ab665cff#the-bytes-layout)

use std::{any::type_name, io, mem::size_of};

use ethereum_types::U256;
use keccak_hash::H256;
use mpt_trie_type_1::nibbles::Nibbles;
use serde::de::DeserializeOwned;
use winnow::{
    combinator::{alt, empty, eof, preceded, repeat_till, trace},
    error::{ErrMode, ErrorKind, FromExternalError, TreeError},
    stream::Stream,
    token::{any, literal, take},
    PResult, Parser as _,
};

#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub struct Witness {
    pub header: V1Header,
    pub instructions: Vec<Instruction>,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub struct V1Header;

#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub enum Instruction {
    Leaf {
        key: Nibbles,
        value: Vec<u8>,
    },
    Extension(Nibbles),
    Branch {
        mask: u32,
    },
    Hash(H256),
    Code(Vec<u8>),
    AccountLeaf {
        key: Nibbles,
        nonce: Option<U256>,
        balance: Option<U256>,
        // according to the spec, this should be `has_code: bool`,
        // but in actuality, there's an additional field here.
        code_size: Option<u64>,
        has_storage: bool,
    },
    NewTrie,
}

/// Parameterise our combinators over the error type to facilitate opt-in
/// debugging.
///
/// Define this trait for conciseness
trait ParserError<'a>:
    winnow::error::ParserError<&'a [u8]>
    + FromExternalError<&'a [u8], ciborium::de::Error<io::Error>>
    + FromExternalError<&'a [u8], UnrecognisedBits>
{
}

impl<'a, E> ParserError<'a> for E where
    E: winnow::error::ParserError<&'a [u8]>
        + FromExternalError<&'a [u8], ciborium::de::Error<io::Error>>
        + FromExternalError<&'a [u8], UnrecognisedBits>
{
}

fn witness<'a, E: ParserError<'a>>(input: &mut &'a [u8]) -> PResult<Witness, E> {
    Ok(Witness {
        header: header.parse_next(input)?,
        instructions: repeat_till(.., instruction, eof)
            .map(|(it, _)| it)
            .parse_next(input)?,
    })
}

fn header<'a, E: ParserError<'a>>(input: &mut &'a [u8]) -> PResult<V1Header, E> {
    trace("header", literal(1).value(V1Header)).parse_next(input)
}

fn instruction<'a, E: ParserError<'a>>(input: &mut &'a [u8]) -> PResult<Instruction, E> {
    let start = input.checkpoint();
    let opcode = any.parse_next(input)?;
    // I don't like the `winnow::combinator::dispatch!` macro
    match opcode {
        0x00 => trace(
            "leaf",
            (key, cbor).map(|(key, value)| Instruction::Leaf { key, value }),
        )
        .parse_next(input),
        0x01 => trace("extension", key.map(Instruction::Extension)).parse_next(input),
        0x02 => trace("branch", cbor.map(|mask| Instruction::Branch { mask })).parse_next(input),
        0x03 => trace("hash", h256.map(Instruction::Hash)).parse_next(input),
        0x04 => trace("code", cbor.map(Instruction::Code)).parse_next(input),
        0x05 => trace("account_leaf", account_leaf).parse_next(input),
        0xBB => trace("new_trie", empty.value(Instruction::NewTrie)).parse_next(input),
        _ => {
            input.reset(&start);
            Err(ErrMode::Backtrack(E::from_error_kind(
                input,
                ErrorKind::Alt,
            )))
        }
    }
}

#[derive(thiserror::Error, Debug)]
#[error("unrecognised bits in flags for account leaf")]
struct UnrecognisedBits;

fn account_leaf<'a, E: ParserError<'a>>(input: &mut &'a [u8]) -> PResult<Instruction, E> {
    bitflags::bitflags! {
        struct AccountLeafFlags: u8 {
            const HAS_CODE = 0b0000_0001;
            const HAS_STORAGE = 0b0000_0010;
            const ENCODES_NONCE = 0b0000_0100;
            const ENCODES_BALANCE = 0b0000_1000;
        }
    }

    let key = key.parse_next(input)?;
    let flags = any
        .try_map(|byte| AccountLeafFlags::from_bits(byte).ok_or(UnrecognisedBits))
        .parse_next(input)?;

    Ok(Instruction::AccountLeaf {
        key,
        nonce: match flags.contains(AccountLeafFlags::ENCODES_NONCE) {
            true => Some(trace("nonce", u256).parse_next(input)?),
            false => None,
        },
        balance: match flags.contains(AccountLeafFlags::ENCODES_BALANCE) {
            true => Some(trace("balance", u256).parse_next(input)?),
            false => None,
        },
        // TODO(0xaatif): brendan's code deviates from the spec and reads a u64 if HAS_CODE
        code_size: match flags.contains(AccountLeafFlags::HAS_CODE) {
            true => Some((trace("code_size", cbor)).parse_next(input)?),
            false => None,
        },
        has_storage: flags.contains(AccountLeafFlags::HAS_STORAGE),
    })
}

fn cbor<'a, T: DeserializeOwned, E: ParserError<'a>>(input: &mut &'a [u8]) -> PResult<T, E> {
    trace(
        format!("cbor{{{}}}", type_name::<T>()),
        |input: &mut &'a [u8]| {
            let start = input.checkpoint();
            match ciborium::from_reader::<T, _>(&mut *input) {
                Ok(it) => Ok(it),
                Err(e) => {
                    input.reset(&start);
                    Err(FromExternalError::from_external_error(
                        input,
                        ErrorKind::Verify,
                        e,
                    ))
                }
            }
        },
    )
    .parse_next(input)
}

fn key<'a, E>(input: &mut &'a [u8]) -> PResult<Nibbles, E>
where
    E: ParserError<'a> + FromExternalError<&'a [u8], ciborium::de::Error<io::Error>>,
{
    trace(
        "key",
        cbor::<Vec<u8>, _>.map(|bytes| {
            let mut key = Nibbles::default();

            if bytes.is_empty() {
                return key;
            }

            // I have no idea why Erigon is doing this with their keys, as I'm don't think
            // this is part of the yellow paper at all?
            if bytes.len() == 1 {
                let low_nib = bytes[0] & 0b00001111;
                key.push_nibble_back(low_nib);
            }

            let flags = bytes[0];
            let is_odd = (flags & 0b00000001) != 0;
            let has_term = (flags & 0b00000010) != 0;

            // ... Term bit seems to have no effect on the key?
            let actual_key_bytes = match has_term {
                false => &bytes[1..],
                true => &bytes[1..],
            };

            if actual_key_bytes.is_empty() {
                return key;
            }

            let final_byte_idx = actual_key_bytes.len() - 1;

            // The compact key format is kind of weird. We need to read the nibbles
            // backwards from how we expect it internally.
            for byte in &actual_key_bytes[..(final_byte_idx)] {
                let high_nib = (byte & 0b11110000) >> 4;
                let low_nib = byte & 0b00001111;

                key.push_nibble_back(high_nib);
                key.push_nibble_back(low_nib);
            }

            // The final byte we might need to ignore the last nibble, so we need to do it
            // separately.
            let final_byte = actual_key_bytes[final_byte_idx];
            let high_nib = (final_byte & 0b11110000) >> 4;
            key.push_nibble_back(high_nib);

            if !is_odd {
                let low_nib = final_byte & 0b00001111;
                key.push_nibble_back(low_nib);
            }

            key
        }),
    )
    .parse_next(input)
}

fn h256<'a, E: ParserError<'a>>(input: &mut &'a [u8]) -> PResult<H256, E> {
    array.map(H256).parse_next(input)
}

fn u256<'a, E: ParserError<'a>>(input: &mut &'a [u8]) -> PResult<U256, E> {
    array::<{ size_of::<U256>() }, _>
        .map(|it| U256::from_big_endian(&it))
        .parse_next(input)
}

fn array<'a, const N: usize, E: ParserError<'a>>(input: &mut &'a [u8]) -> PResult<[u8; N], E> {
    take(N)
        .map(|it: &[u8]| it.try_into().expect("take has already selected N bytes"))
        .parse_next(input)
}

#[cfg(test)]
#[track_caller]
fn do_test<'a, T: PartialEq + core::fmt::Debug>(
    src: &'a [u8],
    expected: T,
    mut parser: impl winnow::Parser<&'a [u8], T, winnow::error::ContextError>,
) {
    let actual = parser.parse(src.as_ref()).unwrap();
    assert_eq!(expected, actual)
}

/// <https://github.com/cbor/test-vectors/blob/aba89b653e484bc8573c22f3ff35641d79dfd8c1/appendix_a.json>
#[test]
fn cbor_test_vectors() {
    do_test(b"\x00", 0, cbor);
    do_test(b"\x01", 1, cbor);
    do_test(b"\x0a", 10, cbor);
    do_test(b"\x17", 23, cbor);
}

#[test]
fn witness_test_vectors() {
    use hex_literal::hex;
    use winnow::error::TreeError;

    let simple = hex!("01004110443132333400411044313233340218300042035044313233350218180158200000000000000000000000000000000000000000000000000000000000000012");
    dbg!(witness::<TreeError<_>>.parse(&simple)).unwrap();
}
