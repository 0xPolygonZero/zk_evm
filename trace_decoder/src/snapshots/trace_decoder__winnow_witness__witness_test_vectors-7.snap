---
source: trace_decoder/src/winnow_witness.rs
expression: "witness::<winnow::error::ContextError>(&mut &*vector.bytes)"
---
Err(
    Backtrack(
        ContextError {
            context: [],
            cause: None,
        },
    ),
)
